### PR TITLE
`accelerate` support for `OwlViT`

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

This PR adds `accelerate` support for `OwlViT` so that any model from this family can be loaded in `8-bit`! Here is a small snippet on how to load and run the model in `8-bit` (based on the snippet from the model card):
```
# pip install accelerate bitsandbytes
import requests
from PIL import Image
import torch

from transformers import OwlViTProcessor, OwlViTForObjectDetection

processor = OwlViTProcessor.from_pretrained("google/owlvit-base-patch32")
model = OwlViTForObjectDetection.from_pretrained("google/owlvit-base-patch32", device_map="auto", load_in_8bit=True)

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)
texts = [["a photo of a cat", "a photo of a dog"]]
inputs = processor(text=texts, images=image, return_tensors="pt")
outputs = model(**inputs)

# Target image sizes (height, width) to rescale box predictions [batch_size, 2]
target_sizes = torch.Tensor([image.size[::-1]])
# Convert outputs (bounding boxes and class logits) to COCO API
results = processor.post_process(outputs=outputs, target_sizes=target_sizes)

i = 0  # Retrieve predictions for the first image for the corresponding text queries
text = texts[i]
boxes, scores, labels = results[i]["boxes"], results[i]["scores"], results[i]["labels"]

# Print detected objects and rescaled box coordinates
score_threshold = 0.1
for box, score, label in zip(boxes, scores, labels):
    box = [round(i, 2) for i in box.tolist()]
    if score >= score_threshold:
        print(f"Detected {text[label]} with confidence {round(score.item(), 3)} at location {box}")
>>> Detected a photo of a cat with confidence 0.705 at location [321.25, 19.8, 643.12, 376.88]
>>> Detected a photo of a cat with confidence 0.729 at location [0.94, 53.55, 319.69, 473.91]
```
Added also a slow test to make sure users can run inference in `fp16` (`8bit` converts the model in. `fp16` under the hood)

All slow tests pass except the one mentioned in #20410 that should be fixed once merged

cc @alaradirik @sgugger @NielsRogge

### CODE REVIEW

**Review by alaradirik — APPROVED** [review]
Thanks for adding this, LGTM!

File: src/transformers/models/owlvit/modeling_owlvit.py [hunk 1]
```diff
@@ -434,6 +434,10 @@ def forward(
 
         attn_probs = nn.functional.dropout(attn_weights, p=self.dropout, training=self.training)
 
+        # For int8 compatibility, sometimes the `attn_probs` are in `fp32`
+        if attn_probs.dtype != value_states.dtype:
+            attn_probs = attn_probs.to(value_states.dtype)
```
Comment (sgugger) [inline src/transformers/models/owlvit/modeling_owlvit.py:439]: No need for the test, PyTorch will do it for you and default to a noop :-)

File: src/transformers/models/owlvit/modeling_owlvit.py [hunk 2]
```diff
@@ -939,8 +945,14 @@ def forward(
         )
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
 
+        # Cast the input to the expected `dtype`
+        expected_input_dtype = self.embeddings.patch_embedding.weight.dtype
+        if expected_input_dtype != pixel_values.dtype:
+            pixel_values = pixel_values.to(expected_input_dtype)
```
Comment (sgugger) [inline src/transformers/models/owlvit/modeling_owlvit.py:951]: Same here

**Review by sgugger — APPROVED** [review]
Thanks for adding this!

File: src/transformers/models/owlvit/modeling_owlvit.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/models/owlvit/modeling_owlvit.py:439]: Ah yes I remember now you pointed me to that last time, will change it

File: src/transformers/models/owlvit/modeling_owlvit.py (hunk 2 above)
Comment (younesbelkada) [reply src/transformers/models/owlvit/modeling_owlvit.py]: Fixed in [57c4271](https://github.com/huggingface/transformers/pull/20411/commits/57c427152f4552666c1e3bb2d5975084c8432ef9)

File: src/transformers/models/owlvit/modeling_owlvit.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/models/owlvit/modeling_owlvit.py]: Fixed in [57c4271](https://github.com/huggingface/transformers/pull/20411/commits/57c427152f4552666c1e3bb2d5975084c8432ef9)

**Review by NielsRogge — APPROVED** [review]
Thanks for working on this!

**Comment by younesbelkada** [comment]
Thanks everyone!

**Pull request merged**